specs/interop/liquidity_migration.md

```diff
@@ -0,0 +1,358 @@
+
+```solidity
+function convertFromSuper(address _legacyAddr, address _superAddr, uint256 _amount) public {
+  require(checkPair(_legacyAddr, _superAddr), "Invalid address pair");
+  require(IERC20(_legacyAddr).decimals() == IERC20(_superAddr).decimals(), "Decimals do not match")
+
+  IERC20(_superAddr).burn(msg.sender, _amount);
+  IERC20(_legacyAddr).mint(msg.sender, _amount);
+
+  emit ConvertedFromSuper(_legacyAddr, _superAddr, msg.sender, _amount);
+}
+```
+
+### Events
+
+#### `ConvertedToSuper`
+
+It MUST trigger when anyone converts legacy tokens
+to `SuperchainERC20` with `convertToSuper`.
+
+```solidity
+event ConvertedToSuper(address indexed legacyAddr, address indexed superAddr, address indexed caller, uint256 amount, uint256 adjustedAmount);
+```
+
+#### `ConvertedFromSuper`
+
+It MUST trigger when anyone converts `SuperchainERC20`
+into legacy tokens with `convertFromSuper`.
+
+```solidity
+event ConvertedFromSuper(address indexed legacyAddr, address indexed superAddr, address indexed caller, uint256 amount);
+```
+
+### Internal Functions
+
+#### `checkPair`
+
+Checks the validity of the `_legacyAddr` and `_superAddr` pair.
+To do so, it will verify:
+
+1. Valid `OptimismMintableERC20` address:
+
+   1. For representations minted before
+      the `OptimismMintableERC20Factory` upgrade:
+      Check an allowed token list that connects L2 representations with the
+      corresponding token address in the native chain (_remote token_ address).
+      The allowed list will live in the `OptimismMintableERC20Factory`
+      and will be a mapping from the legacy address to the _remote token_ address
+      (L1 token address for tokens native to L1).
+   2. For representations minted after
+      the `OptimismMintableERC20Factory` upgrade:
+      Verify the `OptimismMintableERC20Factories` deployment lists.
+      This can be the same mapping used for 1a.
+
+   This approach requires updating
+   the `OptimismMintableERC20Factory` to store
+   deployments as a mapping from `OptimismMintableERC20`
+   to its `REMOTE_TOKEN`
+   (both allowed list and new deployments).
+
+2. Valid `SuperchainERC20` address:
+   checks the `SuperchainERC20Factory` mapping
+   that stores the corresponding remote token address
+   for each `SuperchainERC20` address.
+3. Both tokens correspond to the same remote token address:
+   it will compare the remote token addresses stored in the deployment
+   mappings and revert with `InvalidPair` if they do not match.
+
+You can read more details about access control in the [Appendix](#decimal-management).
+
+A reference implementation would look like follows
+
+```solidity
+IOptimismMintableERC20Factory public factory = IOptimismMintableERC20Factory(Predeploys.OptimismMintableERC20Factory);
+ISuperchainERC20Factory public superFactory = ISuperchainERC20Factory(Predeploys.SuperchainERC20Factory);
+
+function isValidPair(address _legacyAddr, address _superAddr) internal view returns (bool) {
```

Comment:
should this be `checkPair`? or should that be renamed to `isValidPair`?

Reply:
I think this function should also be responsible for checking erc20 metadata such as decimals

Reply:
> should this be `checkPair`? or should that be renamed to `isValidPair`?

Yes, that was a typo (I changed the function name last min). Will fix in the next commit.

> I think this function should also be responsible for checking erc20 metadata such as decimals

Absolutely. We have been keeping functions as specific as possible, but checking decimal compatibility can be considered part of "checking the Pair validity". Will move it.

Reply:
We can also remove the decimal check altogether, wdyt?